src/app/CommandHandler.h

```diff
@@ -549,24 +563,26 @@ class CommandHandler
      *
      * @param [in] aRequestCommandPath the concrete path of the command we are
      *             responding to.
-     * @param [in] aData the data for the response.
+     * @param [in] aResponseCommandPath the concrete path of the command we are
+     *             responding to.
+     * @param [in] encodeCommandDataFunction A lambda function responsible for
+     *             encoding the CommandData field.
      */
-    template <typename CommandData>
-    CHIP_ERROR TryAddResponseData(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
+    template <typename Function>
+    CHIP_ERROR TryAddResponseData(const ConcreteCommandPath & aRequestCommandPath, const ConcreteCommandPath & aResponseCommandPath,
+                                  Function && encodeCommandDataFunction)
     {
         // Return early in case of requests targeted to a group, since they should not add a response.
         VerifyOrReturnValue(!IsGroupRequest(), CHIP_NO_ERROR);
 
         InvokeResponseParameters prepareParams(aRequestCommandPath);
         prepareParams.SetStartOrEndDataStruct(false);
 
-        ConcreteCommandPath responsePath = { aRequestCommandPath.mEndpointId, aRequestCommandPath.mClusterId,
-                                             CommandData::GetCommandId() };
-        ScopedChange internalCallToAddResponse(mInternalCallToAddResponseData, true);
-        ReturnErrorOnFailure(PrepareInvokeResponseCommand(responsePath, prepareParams));
+        ScopedChange<bool> internalCallToAddResponse(mInternalCallToAddResponseData, true);
+        ReturnErrorOnFailure(PrepareInvokeResponseCommand(aResponseCommandPath, prepareParams));
         TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
         VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-        ReturnErrorOnFailure(DataModel::Encode(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields), aData));
+        ReturnErrorOnFailure(encodeCommandDataFunction(*writer));
```

Comment:
So this change is not great; it's forcing all callers to handle the tags themselves, etc...  Is this actually reducing compiled code size?   It's not clear to me why, but I suggest:

1. Documenting in the caller _why_ factoring this way reduces code size.
2. Filing a followup to see whether we can have things better factored without regressing code size.